### DISCUSSION TITLE
MPI failed to bind memory inside hovorod docker

### DISCUSSION THREAD

**Discussion by ghost** [horovod/horovod#2758] [Q&A]
**Environment:**
1. Framework: `tensorflow-gpu==2.1.0`
3. Horovod version: `0.19.3`
4. MPI version: `4.0.3`
5. CUDA version: `10.1`
6. NCCL version: `2.4.8-1+cuda10.1`
7. Python version: `3.6.9`
8. OS and version: `Ubuntu 18.04.3 LTS`
9. GCC version: `gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0`

**Checklist:**
1. [x] Did you search issues to find if somebody asked this question before?
2. [x] If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. [x] If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. [x] Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**

I build from [this Dockerfile](https://github.com/horovod/horovod/blob/master/Dockerfile.gpu), and try to run [BERT pretraining](https://github.com/thorjohnsen/bert/tree/gpu_optimizations) with horovod inside the docker.

After running this command
```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 mpiexec --bind-to socket -np 8 python run_pretraining.py \
--horovod \
--a lot of other args....
```
I get a warning from Open MPI
```
--------------------------------------------------------------------------
WARNING: Open MPI tried to bind a process but failed.  This is a
warning only; your job will continue, though performance may
be degraded.

  Local host:        f643614a5158
  Application name:  /usr/bin/python
  Error message:     failed to bind memory
  Location:          rtc_hwloc.c:447

--------------------------------------------------------------------------
```

Compared to runing the code on the server directly, running in docker with 8 GPUs performs 7~8 times slower. How can I resolve it? Thank you.

**Comment by tgaddair** [comment]
Hey @zhxh1, can you try adding `-map-by socket -rank-by core` when using `-bind-to socket`?

cc @nvcastet who may have more insight here.

**Comment by ghost** [comment]
Thanks for your reply @tgaddair !

The MPI binding error is gone with your solution, but a new error arises:

```
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[b61066046cf1:139385] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[18066,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```

**Comment by nvcastet** [comment]
I don't have much more advice. For containers, those warnings and MPI_Init issues could appear when some linux capabilities are not enabled on the container and also when ulimit max numbers are too low.
